test/robot_test.coffee

```diff
@@ -225,3 +216,85 @@ describe 'Robot', ->
         done()
 
       @robot.receive testMessage
+
+    it 'passes on a message that a prereceive hook allows', (done) ->
+      testMessage = new TextMessage(@user, 'message123')
+      @robot.prereceive (hook) ->
+        hook.message.addedData = "added data"
+      @robot.hear /^message123$/, (response) ->
+        expect(response.message.addedData).to.equal("added data")
+        done()
+      @robot.receive testMessage
+
+    it 'provides a response object so replies can be sent before a match', (done) ->
+      testMessage = new TextMessage(@user, 'message123')
+      replier = @robot.adapter.reply = sinon.spy()
+      listenerCallback = sinon.spy()
+      @robot.hear /^message123$/, listenerCallback
+      @robot.prereceive (hook) ->
+        hook.response.reply "Listen bud, here's why I won't let you do that."
+        hook.finish()
+      @robot.receive testMessage
+      expect(listenerCallback).to.not.have.been.called
+      expect(replier).to.have.been.calledOnce
+      done()
+
+    it 'does not pass on a message that a prereceive hook finishes', (done) ->
+      testMessage = new TextMessage(@user, 'message123')
+      listenerCallback = sinon.spy()
+      @robot.hear /^message123$/, listenerCallback
+      @robot.prereceive (hook) ->
+        hook.finish()
+      @robot.receive testMessage
+      expect(listenerCallback).to.not.have.been.calledOnce
+      done()
+
+    it 'runs a prelistener hook before each listener', (done) ->
+      testMessage = new TextMessage(@user, 'message123')
+      @robot.hear /^message123$/, (response) ->
+        expect(response.message.addedData).to.equal("added data")
+        done()
+      listener = @robot.listeners[0]
+      @robot.prelisten (hook) ->
+        hook.message.addedData = "added data"
+        expect(hook.listener).to.equal(listener)
```

Comment:
I initially had a version where you had to call `hook.next()` instead of optionally calling `hook.finish()`. This was cool because things were still synchronous, but you could make a message wait on an external API call, SMS 2-factor, or whatever else. It was pretty annoying in practice, though, because not calling either in a hook meant *nothing happened.* It was an easy gotcha from a development perspective, and with after scientific survey with zero respondents, I decided the external service case was not common enough to optimize for.

If we want that, though, we can do a better job by raising an error if a hook exits without calling either. Feedback wanted.